src/verso.rs

```diff
@@ -374,67 +394,109 @@ impl Verso {
     ///
     /// - Handle Winit's event, updating Compositor and sending messages to Constellation.
     /// - Handle Servo's messages and updating Compositor again.
-    pub fn run(&mut self, event: Event<()>) {
+    pub fn run(&mut self, event: Event<()>, evl: &EventLoopWindowTarget<()>) {
         self.handle_winit_event(event);
-        self.handle_servo_messages();
+        self.handle_servo_messages(evl);
+        if self.windows.is_empty() {
+            self.compositor
+                .as_mut()
+                .map(IOCompositor::maybe_start_shutting_down);
+        }
     }
 
     /// Handle Winit events
     fn handle_winit_event(&mut self, event: Event<()>) {
         log::trace!("Verso is handling Winit event: {event:?}");
         match event {
-            Event::NewEvents(StartCause::Init) => {
-                // Send the constellation message to start Panel UI
-                let panel_id = self.window.panel.webview_id;
-                let path = self.resource_dir.join("panel.html");
-                let url = ServoUrl::from_file_path(path.to_str().unwrap()).unwrap();
-                send_to_constellation(
-                    &self.constellation_sender,
-                    ConstellationMsg::NewWebView(url, panel_id),
-                );
-            }
-            Event::Suspended | Event::Resumed | Event::UserEvent(()) => {}
-            Event::WindowEvent {
-                window_id: _,
-                event,
-            } => {
+            Event::NewEvents(_) | Event::Suspended | Event::Resumed | Event::UserEvent(()) => {}
+            Event::WindowEvent { window_id, event } => {
                 if let Some(compositor) = &mut self.compositor {
-                    self.window.handle_winit_window_event(
-                        &self.constellation_sender,
-                        compositor,
-                        &event,
-                    )
+                    if let WindowEvent::CloseRequested = event {
+                        // self.windows.remove(&window_id);
+                        compositor.maybe_start_shutting_down();
+                    } else {
+                        let mut need_repaint = false;
+                        for (id, window) in &mut self.windows {
+                            if window_id == *id {
+                                need_repaint = window.handle_winit_window_event(
+                                    &self.constellation_sender,
+                                    compositor,
+                                    &event,
+                                );
+                            }
+                        }
+
+                        if need_repaint {
+                            compositor.repaint_synchronously(&mut self.windows);
+                        }
```

Comment:
```suggestion
let need_repaint = self.windows.iter_mut().any(|(id, window)| {
    window_id == *id && window.handle_winit_window_event(
        &self.constellation_sender,
        compositor,
        &event,
    )
});

if need_repaint {
    compositor.repaint_synchronously(&mut self.windows);
}
```

Reply:
This dosn't seem like the same as I want to only call `handle_winit_window_event` when the ID is correct.